process/block/preprocess/rewardTxPreProcessor.go

```diff
@@ -0,0 +1,454 @@
+	onlyRewardTxsFromOthersBody := block.Body{}
+	for _, mb := range body {
+		if mb.Type != block.RewardsBlock {
+			continue
+		}
+		if mb.SenderShardID == rtp.shardCoordinator.SelfId() {
+			continue
+		}
+
+		onlyRewardTxsFromOthersBody = append(onlyRewardTxsFromOthersBody, mb)
+	}
+
+	missingTxsForShard := rtp.computeExistingAndMissing(
+		onlyRewardTxsFromOthersBody,
+		&rtp.rewardTxsForBlock,
+		rtp.chReceivedAllRewardTxs,
+		block.RewardsBlock,
+		rtp.rewardTxPool,
+	)
+
+	return missingTxsForShard
+}
+
+// processRewardTransaction processes a reward transaction, if the transactions has an error it removes it from pool
+func (rtp *rewardTxPreprocessor) processRewardTransaction(
+	rewardTxHash []byte,
+	rewardTx *rewardTx.RewardTx,
+	round uint64,
+	sndShardId uint32,
+	dstShardId uint32,
+) error {
+
+	err := rtp.rewardsProcessor.ProcessRewardTransaction(rewardTx)
+	if err != nil {
+		return err
+	}
+
+	txShardInfo := &txShardInfo{senderShardID: sndShardId, receiverShardID: dstShardId}
+	rtp.rewardTxsForBlock.mutTxsForBlock.Lock()
+	rtp.rewardTxsForBlock.txHashAndInfo[string(rewardTxHash)] = &txInfo{tx: rewardTx, txShardInfo: txShardInfo}
+	rtp.rewardTxsForBlock.mutTxsForBlock.Unlock()
+
+	return nil
+}
+
+// RequestTransactionsForMiniBlock requests missing reward transactions for a certain miniblock
+func (rtp *rewardTxPreprocessor) RequestTransactionsForMiniBlock(mb block.MiniBlock) int {
+	missingRewardTxsForMiniBlock := rtp.computeMissingRewardTxsForMiniBlock(mb)
+	rtp.onRequestRewardTx(mb.SenderShardID, missingRewardTxsForMiniBlock)
+
+	return len(missingRewardTxsForMiniBlock)
+}
+
+// computeMissingRewardTxsForMiniBlock computes missing reward transactions for a certain miniblock
+func (rtp *rewardTxPreprocessor) computeMissingRewardTxsForMiniBlock(mb block.MiniBlock) [][]byte {
+	missingRewardTxs := make([][]byte, 0)
+	if mb.Type != block.RewardsBlock {
+		return missingRewardTxs
+	}
+
+	for _, txHash := range mb.TxHashes {
+		tx, _ := process.GetTransactionHandlerFromPool(
+			mb.SenderShardID,
+			mb.ReceiverShardID,
+			txHash,
+			rtp.rewardTxPool,
+		)
+
+		if tx == nil {
+			missingRewardTxs = append(missingRewardTxs, txHash)
+		}
+	}
+
+	return missingRewardTxs
+}
+
+// getAllRewardTxsFromMiniBlock gets all the reward transactions from a miniblock into a new structure
+func (rtp *rewardTxPreprocessor) getAllRewardTxsFromMiniBlock(
+	mb *block.MiniBlock,
+	haveTime func() bool,
+) ([]*rewardTx.RewardTx, [][]byte, error) {
+
+	strCache := process.ShardCacherIdentifier(mb.SenderShardID, mb.ReceiverShardID)
+	txCache := rtp.rewardTxPool.ShardDataStore(strCache)
+	if txCache == nil {
+		return nil, nil, process.ErrNilRewardTxDataPool
+	}
+
+	// verify if all reward transactions exists
+	rewardTxs := make([]*rewardTx.RewardTx, 0)
+	txHashes := make([][]byte, 0)
+	for _, txHash := range mb.TxHashes {
+		if !haveTime() {
+			return nil, nil, process.ErrTimeIsOut
+		}
+
+		tmp, _ := txCache.Peek(txHash)
```

Comment:
get the other value, and test that. tmp, ok := txCache.Peek if !ok

Reply:
done